docs/design/ha_master.md

```diff
@@ -0,0 +1,188 @@
+
+**Author:** filipg@, jsz@
+
+# Introduction
+We want to allow users to easily replicate kubernetes masters to have highly available cluster,
+using automated tools like kube-up.sh and kube-down.sh.
+
+This document describes technical design of this feature. It assumes that we are using aforementioned
+scripts for cluster deployment. It focuses on GCE use-case but all of the ideas described in
+the following sections should be easy to port to AWS, other cloud providers and bare-metal environment.
+
+# Overview
+In a cluster with replicated master machine we will have the following component interactions:
+* All etcd replicas will be clustered together and will be using master election
+and quorum mechanism to agree on the state. All of these mechanisms are integral
+parts of etcd and we will only have to configure them properly.
+* All apiserver replicas will be working independently talking to a etcd on the
+same machine, which if needed will forward requests to the current etcd master
+(as explained (here)[https://coreos.com/etcd/docs/latest/getting-started-with-etcd.html]);
+in front of apiservers we will have an L4 load balancer to spread the load among all the replicas.
+* Controller manager, scheduler & cluster autoscaler will use lease mechanism and
+only a single instance will be an active master. All other will be waiting in a standby mode.
+* All add-on managers will work independently and each of them will try to keep add-ons in sync
+
+# Detailed design
+
+## VM naming
+VMs with master replicas will have the following names:
+* first VM - `${MASTER_NAME}`
+* next VMs - `${MASTER_NAME}-<hash>`, where `<hash>` is a 3-characters long alphanumeric string.
+
+It will be ok to delete VM without the suffix.
+
+If user will run deployment scripts with a different `MASTER_NAME` variable than earlier,
+it will fail with an error that there are no masters to replicate.
+
+##Components
+
+###etcd
+```Note: this paragraphs assumes we are using etcd v2.2; it will have to be revisited during
+upgrade to v3, e.g. different command line options.```
+
+All etcd instances will be clustered together and one of them will be an elected master.
+In order to commit any change quorum of the cluster will have to confirm it. Etcd will be
+configured in such a way that all writes and reads will go through the master (requests
+will be forwarded by the local etcd server such that it’s invisible for the user). It will
+affect latency for all operations but it should not increase by much more than the network
+latency between master replicas (latency between zones with a region is < 10ms).
+
+Currently etcd exposes port only using localhost interface. In order to allow clustering
+and inter-VM communication we will also have to use public interface. To secure the
+communication we will configure firewall rules to only allow traffic on etcd port from
+other master machines.
+
+Another option to secure etcd cluster communication would be to use SSL (as described
```

Comment:
The security risk depends on whether you trust the network that the masters are running on. If you don't trust everything running on the network, then you need to use ssl.